### ISSUE TITLE
Pipeline components for preprocessing of images

### ISSUE THREAD

**Issue by kevinyamauchi** [spacetx/starfish#945]
# The issue
Most microscopy images will require some amount of correction before processing to account for systematic errors (e.g.,  drift, illumination non-uniformity, channel crosstalk). Currently, the starfish expects the ingested images to already be corrected and thus the built in logging does not capture the corrects that were performed. Given that these types of corrections can have a significant effect on the downstream processing, I think it is important that these steps are performed in starfish and thus are entered into the logs.

## Example corrections
Some common corrections that I think would be useful to support include:
* Illumination non-uniformity: the illumination intensity will vary across the field of view for most imaging systems. The image must thus be corrected to take this into account so that the observed variations in intensity are due to the sample and not the illumination.
* Channel-to-channel differences in intensity: due to differences in illumination power, the properties of the fluorescent dyes, and tissue autofluorescence, the different fluorescence channels will have different intensities. It is common to normalize image intensities such that they occupy a similar range.
* Channel crosstalk: due to overlapping spectra, signal may bleed over between channels. Sometimes this is intentional to increase multiplexing. Other times this is unavoidable due to the properties of the dyes (e.g., Illumina Miseq sequencing by synthesis chemistry).
* Chromatic shift: a given point in the sample may map to different positions on the camera sensor for different wavelengths (may be due to optical properties or assembly of the imaging system). This is covered by the "Image Registration" pipeline component described in the H1 2019 Roadmap (#895 )
* Stage drift: in multiround imaging, the microscope stage may not return to the exact same spot for a given X,Y coordinate. This is covered by the "Image Registration" pipeline component described in the H1 2019 Roadmap (#895 ).

# Proposed solutions
## Proposed solution 1 (my preferred approach)
We implement two pipeline components that cover the most common corrections:
* Intensity normalization: This would cover both channel-to-channel differences in intensity and illumination non-uniformity. The component would perform element-wise multiplication of a supplied matrix (same size as a `Tile`) with specified `Indices`. For illumination correction, the user would supply the full matrix. For the intensity normalization case, the user could supply a float and the component would create the matrix: `norm_factor * np.ones((tile_shape))`.
* Channel crosstalk: this component performs linear unmixing of channels. See the BaristaSeq notebook.

## Proposed solution 2
We create general pipeline components that perform common matrix math operations to the image tensor and users can compose their own corrections. This seems like it would be less useful to the typical user and the logging may be difficult to interpret.

# Future work
* Learning corrections: It can be possible to learn the appropriate transformation for the corrections listed above. While this is nice, most of them can be directly measured, so I think it is sufficient and potentially more accurate to have the user supply the transformations/coefficients.
* Non-affine transformations: see the discussion in the H1 2019 Roadmap

**Comment by kevinyamauchi** [comment]
@ambrosejcarr and @neuromusic , I've made the pipeline components we discussed in the PRs above (#983, #987, #990). Would you mind taking a look and seeing if they capture the features we discussed? It would be helpful to make sure the necessary use cases are covered (especially with regards to the roadmap). In parallel, I will work on adding some more tests.

**Comment by ambrosejcarr** [comment]
ElementWiseMult and LinearUnmixing got merged. 🎆 

I'm going to close this issue and, we can continue with registration discussions in #1051 @kevinyamauchi let me know if you think there is anything outstanding that's not covered in #1051
